### ISSUE TITLE
API Proposal: instansearch instance events to indicate the DOM is ready

### ISSUE THREAD

**Issue by bobylito** [algolia/instantsearch.js#167]
## The problem

When doing the rendering of all the components we see different manipulations of the dom until the initial search is done. This makes the display flickers and doesn't provide a very polished effect.

## Some thoughts

We could have the components to only display themselves when they are ready to be used (eg : after init for the search input, or after the rendering after the first search for the hits). Because we chose to keep our rendering to the minimum some elements will be displayed that are depending on the integration, therefore we should make it possible to listen to the lifecycle of the instantsearch.js (who said events ;) )

## API proposal

basically the instantsearch.js instance would be an event emitter with two events for the moment  : 
  - `render` which will be triggered after the each rendering of all the widgets that implement `render`

**Comment by vvo** [comment]
Should we have something like a `ready` alias to `afterFirstRender`? Only for ease of use.

**Comment by vvo** [comment]
renamed to "API Proposal: instansearch instance events to indicate the DOM is ready" since fadein fadeout was not the real proposal. Only an effect of it?

**Comment by redox** [comment]
Like the idead @bobylito, it's true that it can flicker. Maybe something like `initialized` instead of `afterInit`?

**Comment by bobylito** [comment]
What I like with my proposal of names is that they relate to the widget lifecycle.

**Comment by pixelastic** [comment]
> We could have the components to only display themselves when they are ready to be used

I'm not sure about that part. I agree that we should not have any flickering of UI, but not sure delaying the display until everything is ready is the way to go. I did not experience the widget flickering, could you provide an example?

**Comment by vvo** [comment]
> I'm not sure about that part. I agree that we should not have any flickering of UI, but not sure delaying the display until everything is ready is the way to go.

With events as proposed it will be up to the user to do it or not (display none all, then show with or without css transition).

> I did not experience the widget flickering, could you provide an example?

Right now the widgets are empty containers that are filled later on when everything is ready. So you can see the border flickering:
- first load, has empty containers with borders
- instantsearch.js loads an init, widgets are filled up

I think if you throttle your connection in chrome you should see it.

**Comment by pixelastic** [comment]
Ok, I understand. I'm in favor of an `afterInit` to stay consistent with the vocabulary we are already using in the widgets.

I'm not so sure about the `afterFirstRender`, some widgets might not render while other are rendered (because we can set the `hideIfNoResult`-like attributes to individual widgets).

**Comment by bobylito** [comment]
We could just have a `afterRender` since we're likely to use event emitters and they support `once`. This could be : 

```js
instantsearch().once('afterRender', fadeInUISomeStuff);
```

**Comment by vvo** [comment]
> Ok, I understand. I'm in favor of an afterInit to stay consistent with the vocabulary we are already using in the widgets.

Careful here we should think about the event name from a user POV not from an implementer POV. So maybe most user will be happy with a `ready` event?

**Comment by bobylito** [comment]
`ready` is meaningless in our case. Its semantic comes from the DOM, which is irrelevant here. Maybe if we have something like afterResults or afterSearch? (for `afterRender`)

Those are advanced feature anyway, and they will require some user willingness to learn the concepts of the library. Let's make it so we don't have the user learn too much by reusing the same concepts.
